### PR TITLE
Mono/Differential switchable, local OTA update, MDNS, playing of single files

### PR DESCRIPTION
**Pull request by warhog**
Some small improvements:
- Made mono/differential switchable by define
- Added local ota update (moved update to separate page)
- Added MDNS service (default is: http://rfidshelf.local)
- Added ability to play single files (a bit hacky)
- Added basic bass & treble control using compile time variables (see TREBLE_AMPLITUDE, TREBLE_FREQLIMIT, BASS_AMPLITUDE, BASS_FREQLIMIT)

### CODE REVIEW

**Comment by TheNitek** [comment]
Wow, thanks for this pull request!
Would you mind adding a little bit more information?
Did the differential output work for you? (I deactivated it because it didn't do anything for me)
What's your experience with the bass/treble values? Did you use the speaker I recommended or a different one?

MDNS and bass/treble look like no brainers to merge. The OTA update doesn't really fit my vision of the URL layout (= GET request reflect the SD cards folder structure) and I guess I have to dig a bit deeper into the code. The later also applies the play file button.

Since I'm currently refactoring the firmware anyway (splitting up into mutiple files), I'll probably have to carefully pick the changes (and/or copy&paste them into my dev version), but I definitely want to integrated them!

**Comment by warhog** [comment]
I have the differential output enabled because it sounds a bit better (at least it feels so :) ). But I have a bit different setup, instead of the PAM8302 I use a MAX98306 which has stereo capability and drives two speakers. The original speakers where not available anymore but I have a pair that looks exactly the same and has the same values (I think they are from the same vendor). I've built everything into an Ikea Dragan box, looks a bit like a Hörbert then.
Enabling the bass made it sound much better. Its not that it is a hifi system now but the overall sound is more balanced. But as this is a highly personal opinion my plan was to integrate it to the web ui (same for differential). Then everyone can change it easily according to his needs/setup.

I have expected the issue with the URL layout :) Although I like the idea of the current setup it makes things more complicated with more and more functionality (just to note one is the huge if condition in the handleNotFound() function or different endpoints for the bass/trebble ui functionality or further ui stuff that is fetch with GET calls or websockets). From my personal experience it would be better to put everything under its own url endpoint (e.g. /sd/<dir> to play files, / for index, /ota, ...). The current layout can be moved to /sd for example to have the folder structure available there. Then you can add an almost infinite number of endpoints without getting into trouble or using huge if conditions.
But that may be just my personal point of view which doesn't has to be the right one :)

Splitting it up into multiple files is a good idea, I wanted to start this but then I'm waiting.

**Comment by TheNitek** [comment]
Well I am not sure yet about the URL either, but I wouldn't want to mix both concepts - I guess that needs some thinking ;-)

I just pushed the split version. It's far from perfect, but it is better than one big file. I am still thinking about how to handle the HTML/JS code, because a mess right now.

Would it be possible to rebase and split this pull request into multiple ones (one per feature)?

**Comment by warhog** [comment]
Yes, in the end maybe it is just a personal preference :)

For HTML/JS code I put them in a char * inline or in an external file if they are bigger. Haven't found an easy and/or better way also.

I can rebase and split into multiple branches but will take some time because I'm away for some days. Or maybe you cherry pick manually bass/treble/differential and put the rest by side for some time til the URL stuff has a final concept.

**Comment by TheNitek** [comment]
Never tried cherry picking from a fork, but sounds fun. I'll try that!

**Comment by TheNitek** [comment]
Wow, bass & treble control is awesome! Thanks so much for that

File: RfidShelf/RfidShelf.ino [hunk 1]
```diff
@@ -499,6 +500,20 @@ void playFile() {
   }
 }
 
+void playFilename(String filepath) {
+  // IO takes time, reset watchdog timer so it does not kill us
+  ESP.wdtFeed();
+  Serial.print(F("Playing "));
+  Serial.println(filepath);
+  playing = PLAYING_FILE;
+  currentFile = filepath;
```
Comment (TheNitek) [inline RfidShelf/RfidShelf.ino]: This doesn't work, because currentFile contains the path now, not only the filename

**Pull request closed without merging**